tree_unique_args/src/type_analysis.egg

```diff
@@ -271,4 +271,29 @@
        (HasType id out ty)
       )
       ((HasType ctx (Let id in out) ty))
-      :ruleset type-analysis)
\ No newline at end of file
+      :ruleset type-analysis)
+
+; Loops
+(rule ((HasTypeDemand ctx (Loop id in pred-out)))
+      ((HasTypeDemand ctx in))
+      :ruleset type-analysis)
+
+(rule (
+       (HasTypeDemand ctx (Loop id in pred-out))
+       (HasType ctx in ty)
+      )
+      (
+       (HasType id (Arg id) ty) ; assert the argument has type ty in the loop's context
+       (HasTypeDemand id pred-out) ; demand the types of the predicate and output in the loop's context
+      )
+      :ruleset type-analysis)
+
+(rule (
```

Comment:
This rule is good, but it doesn't give an error when the `pred-out` doesn't match the tuple format.
So I guess we need another rule?

Reply:
That'll probably get covered later when I add the rule that everything must get a type?
Or I can add an error rule here.

Reply:
It's true that you'll get an error that it didn't get a type...
I guess it'll be a pretty unhelpful error though. "Didn't get a type" should mean there's a bug in the type checker, while other errors mean the code doesn't typecheck